### PR TITLE
Add get-a-repository-README endpoint

### PR DESCRIPTION
**Pull request by onukura**
### Before submitting a pull-request to GitBucket I have first:

- [x] read the [contribution guidelines](https://github.com/gitbucket/gitbucket/blob/master/.github/CONTRIBUTING.md)
- [x] rebased my branch over master
- [x] verified that project is compiling
- [ ] verified that tests are passing
- [x] squashed my commits as appropriate *(keep several commits if it is relevant to understand the PR)*
- [ ] [marked as closed using commit message](https://help.github.com/articles/closing-issues-via-commit-messages/) all issue ID that this PR should correct

## About this PR

This PR add an "Get a repository README" [API endpoint](https://developer.github.com/v3/repos/contents/#get-a-repository-readme).

### CODE REVIEW

File: src/main/scala/gitbucket/core/controller/api/ApiRepositoryContentsControllerBase.scala [hunk 1]
```diff
@@ -13,10 +13,13 @@ import scala.util.Using
 trait ApiRepositoryContentsControllerBase extends ControllerBase {
   self: ReferrerAuthenticator with WritableUsersAuthenticator with RepositoryCommitFileService =>
 
-  /*
+  /**
    * i. Get the README
    * https://developer.github.com/v3/repos/contents/#get-the-readme
```
Comment (SIkebe) [inline src/main/scala/gitbucket/core/controller/api/ApiRepositoryContentsControllerBase.scala:18]:
```suggestion
   * i. Get a repository README
   * https://docs.github.com/en/rest/reference/repos#get-a-repository-readme
```

File: src/main/scala/gitbucket/core/controller/api/ApiRepositoryContentsControllerBase.scala [hunk 2]
```diff
@@ -13,10 +13,13 @@ import scala.util.Using
...
    */
+  get("/api/v3/repos/:owner/:repository/readme")(referrersOnly { repository =>
+    getContents(repository, "README.md", params.getOrElse("ref", repository.repository.defaultBranch))
```
Comment (SIkebe) [inline src/main/scala/gitbucket/core/controller/api/ApiRepositoryContentsControllerBase.scala:21]: This implementation does not work if the repository has a "readme.md" or "README.MD", ...
GitHub allows case-insensitive README file.

Comment (onukura) [reply src/main/scala/gitbucket/core/controller/api/ApiRepositoryContentsControllerBase.scala]: @SIkebe  Thank you for your review!
I add option to `getContents` for matching filename by low-case. https://github.com/gitbucket/gitbucket/pull/2535/commits/10d97a43f9760496ed1a73c580afbe2d4ee10c2b

File: src/main/scala/gitbucket/core/controller/api/ApiRepositoryContentsControllerBase.scala (hunk 1 above)
Comment (onukura) [reply src/main/scala/gitbucket/core/controller/api/ApiRepositoryContentsControllerBase.scala]: fixed in https://github.com/gitbucket/gitbucket/pull/2535/commits/10d97a43f9760496ed1a73c580afbe2d4ee10c2b

File: src/main/scala/gitbucket/core/controller/api/ApiRepositoryContentsControllerBase.scala [hunk 3]
```diff
@@ -34,22 +37,32 @@ trait ApiRepositoryContentsControllerBase extends ControllerBase {
...
 
-  private def getContents(repository: RepositoryService.RepositoryInfo, path: String, refStr: String) = {
-    def getFileInfo(git: Git, revision: String, pathStr: String): Option[FileInfo] = {
+  private def getContents(
+    repository: RepositoryService.RepositoryInfo,
+    path: String,
+    refStr: String,
+    matchByLowCase: Boolean = false
+  ) = {
+    def getFileInfo(git: Git, revision: String, pathStr: String, matchByLowCase: Boolean): Option[FileInfo] = {
```
Comment (SIkebe) [inline src/main/scala/gitbucket/core/controller/api/ApiRepositoryContentsControllerBase.scala:46]: IMHO, `ignorecase` would be better.

Comment (onukura) [reply src/main/scala/gitbucket/core/controller/api/ApiRepositoryContentsControllerBase.scala]: yeah... true... I fixed arg name in https://github.com/gitbucket/gitbucket/pull/2535/commits/f346ad9b55ea83eb93f4a25b43b67fa422949d42
Thanks!

File: src/main/scala/gitbucket/core/controller/api/ApiRepositoryContentsControllerBase.scala [hunk 4]
```diff
@@ -13,10 +13,13 @@ import scala.util.Using
...
 
-  /*
-   * i. Get the README
-   * https://developer.github.com/v3/repos/contents/#get-the-readme
+  /**
+   * i. Get a repository README
+   * https://docs.github.com/en/rest/reference/repos#get-a-repository-readme
    */
+  get("/api/v3/repos/:owner/:repository/readme")(referrersOnly { repository =>
+    getContents(repository, "readme.md", params.getOrElse("ref", repository.repository.defaultBranch), true)
```
Comment (takezoe) [inline src/main/scala/gitbucket/core/controller/api/ApiRepositoryContentsControllerBase.scala:21]: README is not only `readme.md`.
https://github.com/gitbucket/gitbucket/blob/7544f64c65c8d59184ea65dab482c7ba459714c0/src/main/scala/gitbucket/core/controller/RepositoryViewerController.scala#L911-L913

Comment (SIkebe) [reply src/main/scala/gitbucket/core/controller/api/ApiRepositoryContentsControllerBase.scala:21]: Oh, I completely forgot that...

Comment (onukura) [reply src/main/scala/gitbucket/core/controller/api/ApiRepositoryContentsControllerBase.scala:21]: @takezoe Ah... So must fix this PR.

I wonder if there are more than two readme files with different extension, what will happen??

Comment (SIkebe) [reply src/main/scala/gitbucket/core/controller/api/ApiRepositoryContentsControllerBase.scala:21]: I can't see any sorting or priority. The first one in the `readmeFiles`is rendered anyway?
In the below case, I verified that readme.pdf -> readme.md -> readme.txt

image

Comment (SIkebe) [reply src/main/scala/gitbucket/core/controller/api/ApiRepositoryContentsControllerBase.scala:21]: It doesn't matter with this PR, I found a strange behavior. (maybe gitbucket-html5media-plugin's issue?)

matryoshka

Comment (onukura) [reply src/main/scala/gitbucket/core/controller/api/ApiRepositoryContentsControllerBase.scala]: @takezoe fixed PR at https://github.com/gitbucket/gitbucket/pull/2535/commits/54baa5d8bfa0c6f0e1416c1ee6ef3571557ee016. Now this api process readme with any extensions which can be rendered. 
At the same time `RepositoryViewerController.readmeFiles` moved to `RepositoryCommitFileService.readmeFiles` in order to user this List from `ApiRepositoryContentsControllerBase`.

Comment (takezoe) [reply src/main/scala/gitbucket/core/controller/api/ApiRepositoryContentsControllerBase.scala]: Thanks for fixing the pull request!

> I can't see any sorting or priority. The first one in the readmeFilesis rendered anyway?

Having explicit order would be better. We can consider it in another issue.

**Comment by takezoe** [comment]
@onukura Merged. This API also has to be added to https://github.com/gitbucket/gitbucket/wiki/API-WebHook.

**Comment by onukura** [comment]
@takezoe wiki updated.

**Comment by takezoe** [comment]
Thanks!

**Pull request merged**